django_auth_ldap/backend.py

```diff
@@ -45,94 +45,146 @@
 Additional classes can be found in the config module next to this one.
 """
 
+from __future__ import annotations
+
 import copy
 import operator
 import pprint
 import re
 import warnings
 from functools import reduce
+from logging import Logger
+from typing import (
+    AbstractSet,
+    Any,
+    Callable,
+    Collection,
+    Dict,
+    FrozenSet,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
 
 import django.conf
 import django.dispatch
 import ldap
 from django.contrib.auth import get_user_model
-from django.contrib.auth.models import Group, Permission
+from django.contrib.auth.models import AbstractUser, Group, Permission
 from django.core.cache import cache
 from django.core.exceptions import ImproperlyConfigured, ObjectDoesNotExist
+from django.http import HttpRequest
 from django.utils.inspect import func_supports_parameter
+from ldap.ldapobject import LDAPObject
 
 from django_auth_ldap.config import (
+    AbstractLDAPSearch,
     ConfigurationWarning,
     LDAPGroupQuery,
+    LDAPGroupType,
     LDAPSearch,
     _LDAPConfig,
 )
 
-logger = _LDAPConfig.get_logger()
+try:
+    from django.contrib.auth.backends import BaseBackend
+except ImportError:
+    BaseBackend = object
```

Comment:
Why guard against import errors? We can import from django above, `BaseBackend` should always be importable.

Reply:
Since the `BaseBackend` was first introduced in [Django 3.0](https://docs.djangoproject.com/en/dev/releases/3.0/#django-contrib-auth) and `django-auth-ldap` supports Django 2.2, this would result in an `ImportError` otherwise.

Reply:
I see, thanks for explaining. This can definitely use a comment, so that the import error catching is removed when support for Django 2.2 is removed.

Reply:
In form of a TODO or just a simple comment?

Reply:
Both work, just something we can find when grepping `2.2`.

For example:
```python
# Stop catching ImportError when dropping support for Django 2.2.
```